VariantShredding.md

```diff
@@ -25,276 +25,299 @@
-| {a: 123, b: {} | null | null | null | 123 | null | null | b.object.c set to null, to indicate that c is missing |
-| {a: 123, d: 456} | {d: 456} | null | 123 | null | null | null | Extra field d is stored as variant_value |
-| [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | null | null | null | null | null | Not an object |
+The series of measurements `34, null, "n/a", 100` would be stored as:
 
-# Parquet Layout
+| Value   | `metadata`       | `value`               | `typed_value` |
+|---------|------------------|-----------------------|---------------|
+| 34      | `01 00` v1/empty | null                  | `34`          |
+| null    | `01 00` v1/empty | `00` (null)           | null          |
+| "n/a"   | `01 00` v1/empty | `13 6E 2F 61` (`n/a`) | null          |
+| 100     | `01 00` v1/empty | null                  | `100`         |
 
-The `array` and `object` fields represent Variant array and object types, respectively.
-Arrays must use the three-level list structure described in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md.
+Both `value` and `typed_value` are optional fields used together to encode a single value.
+Values in the two fields must be interpreted according to the following table:
 
-An `object` field must be a group.
-Each field name of this inner group corresponds to the Variant value's object field name.
-Each inner field's type is a recursively shredded variant value: that is, the fields of each object field must be one or more of `object`, `array`, `typed_value` or `variant_value`.
+| `value`  | `typed_value` | Meaning                                                  |
+|----------|---------------|----------------------------------------------------------|
+| null     | null          | The value is missing                                     |
+| non-null | null          | The value is present and may be any type, including null |
+| null     | non-null      | The value is present and is the shredded type            |
+| non-null | non-null      | The value is present and a partially shredded object     |
 
-Similarly the elements of an `array` must be a group containing one or more of `object`, `array`, `typed_value` or `variant_value`.
+An object is _partially shredded_ when the `value` is an object and the `typed_value` is a shredded object.
 
-Each leaf in the schema can store an arbitrary Variant value.
-It contains an `variant_value` binary field and a `typed_value` field.
-If non-null, `variant_value` represents the value stored as a Variant binary.
-The `typed_value` field may be any type that has a corresponding Variant type.
-For each value in the data, at most one of the `typed_value` and `variant_value` may be non-null.
-A writer may omit either field, which is equivalent to all rows being null.
+If both fields are non-null and either is not an object, the value is invalid. Readers must either fail or return the `typed_value`.
```

Comment:
@RussellSpitzer and @gene-db, this could use some attention.

Here, if both `value` and `typed_value` are non-null I initially thought it made more sense to prefer `value` because it doesn't need to be re-encoded.

However, this conflicts with object fields, where the value of `typed_value` is preferred so that data skipping is correct. If the object's `value` could contains a field that conflicts with a sub-field's `typed_value` there is no way of knowing from field stats. If we preferred the field value stored in the object's `value` then data skipping could be out of sync with the value returned in the case of a conflict.

Reply:
the `value` is invalid

```suggestion
If both fields are non-null and either is not an object, the `value` is invalid. Readers must either fail or return the `typed_value`.
```

Reply:
The problem is that readers won't actually implement restrictions like this and we can't fully prevent it. It is invalid for a writer to produce a value where `value` and `typed_value` conflict. But writer bugs happen and readers need to know what to do when they encounter that situation. Otherwise we would get different behaviors between readers that are processing the same data file.

It all comes down to end users -- if a writer bug produces data like this, readers _will_ implement the ability to read because the data still exists and can be recovered. When that happens, we want to know how it is interpreted.

Reply:
My take is if readers have bugs and produce invalid values, I'm not sure you can really trust most of the data at all (even metadata).  It sounds like we are assuming 1 specific type of bug where readers accidentally forget to clear a field?

Reply:
The point is not the bug. It is that we want to make it valid to read a projection without checking the `value` for bugs.

Reply:
I don't think I understand. If `a` and `b` are shredded, then there should be no fields with that name in `value` and the spec is stating that you don't need to check for them. That means all of the information needed to continue is in the shredded Parquet columns. That's not specific to Spark.

Reply:
What I was trying to say is, if I am casting that field to a struct with "a" and "b" but "c" exists in value, then some engines will require that cast to fail, because the object with keys "a", "b" and "c" do not match the struct schema. Apparently, with spark such a cast succeeds by producing the struct with the subset of the keys that match the struct schema.